src/transformers/generation_utils.py

```diff
@@ -1253,17 +1269,43 @@ def greedy_search(
             input_ids, max_length
         )
 
+        this_peer_finished = False
         while cur_len < max_length:
+
+            if synced_gpus:
+                # Under synced_gpus the `forward` call must continue until all gpus complete their sequence.
+                # The following logic allows an early break if all peers finished generating their sequence
+                this_peer_finished_flag = torch.tensor(0.0 if this_peer_finished else 1.0).to(input_ids.device)
+                # send 0.0 if we finished, 1.0 otherwise
+                dist.all_reduce(this_peer_finished_flag, op=dist.ReduceOp.SUM)
+                #print(f"{dist.get_rank()} Finished? {this_peer_finished_flag}")
+                # did all peers finish? the reduced sum will be 0.0 then
+                if this_peer_finished_flag.item() == 0.0:
+                    #print(f"{dist.get_rank()} All peers finished!")
+                    break
+            elif this_peer_finished:
+                break
+
+            # if this_peer_finished:
+            #     break
+
```

Comment:
@patrickvonplaten, this is the updated logic that now includes an additional sync between all gpus to allow for breaking early and not waste cycles running till max_length. Please have a look. It's this and a few more blocks below - all inside the existing while loop now.

Once you feel it's good I will port it to the other 4 functions.

If you think it's ok to refactor some of this - at least the syncing code please let me know - it'd make the code easier to read.

Thanks